include/jemalloc/internal/hpa.h

```diff
@@ -0,0 +1,92 @@
+#ifndef JEMALLOC_INTERNAL_HPA_H
+#define JEMALLOC_INTERNAL_HPA_H
+
+#include "jemalloc/internal/geom_grow.h"
+#include "jemalloc/internal/hpa_central.h"
+#include "jemalloc/internal/pai.h"
+#include "jemalloc/internal/psset.h"
+
+typedef struct hpa_s hpa_t;
+struct hpa_s {
+	/*
+	 * We have to mutexes for the central allocator; mtx protects its state,
+	 * while grow_mtx protects controls the ability to grow the backing
+	 * store.  This prevents race conditions in which the central allocator
+	 * has exhausted its memory while mutiple threads are trying to
+	 * allocate.  If they all reserved more address space from the OS
+	 * without synchronization, we'd end consuming much more than necessary.
+	 */
+	malloc_mutex_t grow_mtx;
+	malloc_mutex_t mtx;
+	hpa_central_t central;
+	/* The arena ind we're associated with. */
```

Comment:
Does the central allocator also has an associated arena?

Reply:
It depends on what you mean; it's edatas have a value in those bits that they get from their edata cache, but practically it serves mostly as a source for edata_ts. The arena index of edata_ts in the central arena isn't particularly meaningful (eventually it might be, if we decide to shard it better).